internal/backend/triggers/triggers.go

```diff
@@ -3,106 +3,114 @@ package triggers
-		if schedule, _ := trigger.Data()[fixtures.ScheduleExpression].ToString(); schedule != "" {
-			if err := m.scheduler.Create(ctx, trigger.ID().String(), schedule, trigger.ID()); err != nil {
-				return sdktypes.InvalidTriggerID, err
-			}
+	sl := m.sl.With("trigger_id", trigger.ID())
+
+	switch trigger.SourceType() {
+	case sdktypes.TriggerSourceTypeWebhook:
+		trigger = webhookssvc.InitTrigger(trigger)
+		sl.With("slug", trigger.WebhookSlug()).Infof("creating webhook trigger with slug %q", trigger.WebhookSlug())
+	case sdktypes.TriggerSourceTypeSchedule:
+		if err := m.scheduler.Create(ctx, trigger.ID(), trigger.Schedule()); err != nil {
+			return sdktypes.InvalidTriggerID, fmt.Errorf("create schedule: %w", err)
 		}
+		sl.With("schedule", trigger.Schedule()).Infof("creating schedule trigger with spec %q", trigger.Schedule())
+	case sdktypes.TriggerSourceTypeConnection:
+		sl.With("connection", trigger.ConnectionID()).Infof("creating connection trigger with connection %q", trigger.ConnectionID())
+	default:
+		return sdktypes.InvalidTriggerID, sdkerrors.NewInvalidArgumentError("unsupported source type")
 	}
 
-	// FIXME: see atomicity NOTE above
 	if err := m.db.CreateTrigger(ctx, trigger); err != nil {
 		return sdktypes.InvalidTriggerID, err
 	}
 
 	return trigger.ID(), nil
 }
 
-func (m *triggers) Update(ctx context.Context, trigger sdktypes.Trigger) error {
-	triggerID := trigger.ID()
-	prevTrigger, err := m.db.GetTrigger(ctx, triggerID)
+func (m *triggers) Update(ctx context.Context, next sdktypes.Trigger) error {
+	next = next.WithWebhookSlug("") // ignore webhook slug.
+
+	if next.Equal(sdktypes.InvalidTrigger) {
+		// no update
+		return nil
+	}
+
+	triggerID := next.ID()
+	curr, err := m.db.GetTriggerByID(ctx, triggerID)
 	if err != nil {
 		return err
 	}
 
-	prevData := prevTrigger.Data()
-	prevScheduleVal, isSchedulerPrevTrigger := prevData[fixtures.ScheduleExpression]
-	isSchedulerPrevTrigger = isSchedulerPrevTrigger && prevTrigger.ConnectionID() == sdktypes.BuiltinSchedulerConnectionID
-
-	data := trigger.Data()
-	scheduleVal, isSchedulerTrigger := data[fixtures.ScheduleExpression]
-	schedule, _ := scheduleVal.ToString()
-	isSchedulerTrigger = isSchedulerTrigger && trigger.ConnectionID() == sdktypes.BuiltinSchedulerConnectionID
-
-	if isSchedulerPrevTrigger {
-		if !isSchedulerTrigger { // scheduler trigger -> trigger
-			err = m.scheduler.Delete(ctx, triggerID.String())
-		} else { // scheduler trigger -> scheduler trigger
-			if schedule != prevScheduleVal.String() { // schedule changed
-				err = m.scheduler.Update(ctx, triggerID.String(), schedule)
-			}
-		}
-	} else if isSchedulerTrigger { // trigger -> scheduler trigger
-		err = m.scheduler.Create(ctx, triggerID.String(), schedule, triggerID)
+	if next.Equal(curr) {
+		// no update
+		return nil
 	}
 
-	if err != nil {
-		return err
+	if curr.SourceType() != next.SourceType() {
+		return sdkerrors.NewInvalidArgumentError("cannot update source type")
```

Comment:
just FYI - before your change user was able to update trigger type, which is IMO much more convenient when redefining trigger in the manifest. Otherwise user will be forced to rename it.
I would suggest somehow to support it - either here in update, or in maifest by deleting curr/adding new next if source type is changing.

Reply:
i don't think changing trigger type is something we need to support. worst case a trigger can be deleted and added.